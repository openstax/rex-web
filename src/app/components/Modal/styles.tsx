import React from 'react';
import styled from 'styled-components/macro';
import { toolbarIconColor } from '../../content/components/constants';
import { toolbarIconStyles } from '../../content/components/Toolbar/iconStyles';
import theme from '../../theme';
import Times from '../Times';
import { bodyCopyRegularStyle, h3Style, h4Style } from '../Typography';

export const modalPadding = 3.0;

// tslint:disable-next-line:variable-name
export const Card = styled.div`
  display: flex;
  flex-direction: column;
  margin: auto;
  overflow: hidden;
  width: 40rem;
  ${bodyCopyRegularStyle};
  background-color: white;
  box-shadow: 0 0 2rem rgba(0, 0, 0, 0.05), 0 0 4rem rgba(0, 0, 0, 0.08);
`;

// tslint:disable-next-line:variable-name
export const Header = styled.header`
  display: flex;
  align-items: center;
  margin-bottom: ${modalPadding * 0.5}rem;
  padding: ${modalPadding * 0.5}rem ${modalPadding}rem;
  background: ${theme.color.neutral.pageBackground};
  border-bottom: solid 0.1rem ${theme.color.neutral.darker};
  justify-content: space-between;
`;

// tslint:disable-next-line:variable-name
export const Heading = styled.h1`
  ${h4Style}
  display: flex;
  align-items: center;
  margin: 0;
  padding: ${modalPadding * 0.5}rem 0;
`;

// tslint:disable-next-line:variable-name
export const BodyHeading = styled.h3`
  ${h3Style}
  font-weight: 400;
  padding: ${modalPadding * 0.5}rem 0;
`;

// tslint:disable-next-line:variable-name
export const Body = styled.div`
  display: flex;
  flex-direction: column;
  padding: 0 ${modalPadding}rem;

  ${Card.Header} + & { /* stylelint-disable */
    margin-top: 0;
  }
`;

// tslint:disable-next-line:variable-name
export const Mask = styled.div`
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  position: fixed;
  background-color: rgba(0, 0, 0, 0.3);
`;

// tslint:disable-next-line:variable-name
export const ModalWrapper = styled.div`
  top: 0;
  z-index: ${theme.zIndex.errorPopup};
  left: 0;
  width: 100%;
  height: 100%;
  display: flex;
  position: fixed;
  justify-content: center;
  align-items: center;
`;

// tslint:disable-next-line:variable-name
export const CardWrapper = styled.div`
  z-index: 1;
`;

// tslint:disable-next-line:variable-name
export const CloseModalIcon = styled((props) => <Times {...props} aria-hidden='true' focusable='false' />)`
  ${toolbarIconStyles};
  cursor: pointer;
  margin-right: 0;
  padding-right: 0;
  color: ${toolbarIconColor.lighter};

  :hover {
    color: ${toolbarIconColor.base};
  }
  height: 2rem;
  width: 2rem;
`;

// tslint:disable-next-line:variable-name
export const Footer = styled.div`
  display: flex;
  justify-content: space-between;
  padding: ${modalPadding}rem;
`;
