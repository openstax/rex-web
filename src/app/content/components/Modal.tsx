import React from 'react';
import { createPortal } from 'react-dom';
import { StyledComponentProps } from 'styled-components/macro';
import ScrollLock from '../../components/ScrollLock';
import { assertDocument } from '../../utils/browser-assertions';
import { Modal as ModalWrapper, PopupWrapper } from '../styles/PopupStyles';

interface ModalWithScrollLockProps extends StyledComponentProps<'div', {}, {}, ''> {
  scrollLockProps: React.ComponentProps<typeof ScrollLock>;
}

// tslint:disable-next-line: variable-name
const ModalWithScrollLock = React.forwardRef<HTMLDivElement, ModalWithScrollLockProps>(
  ({ children, scrollLockProps, ...props }: ModalWithScrollLockProps, ref) => {
  return createPortal(
    <PopupWrapper>
      <ScrollLock {...scrollLockProps} />
      <ModalWrapper ref={ref} {...props}>
        {children}
      </ModalWrapper>
    </PopupWrapper>,
    assertDocument().body
  );
});

export default ModalWithScrollLock;
