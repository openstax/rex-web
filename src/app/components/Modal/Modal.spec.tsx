import React from 'react';
import renderer from 'react-test-renderer';
import Modal from '.';
import TestContainer from '../../../test/TestContainer';

describe('Modal', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<TestContainer>
      <Modal
        onModalClose={() => null}
        heading='i18n:discard:heading'
      />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with children', () => {
    const component = renderer.create(<TestContainer>
      <Modal
        onModalClose={() => null}
        heading='i18n:discard:heading'
      >
        <div>child 1</div>
        <div>child 2</div>
      </Modal>
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
