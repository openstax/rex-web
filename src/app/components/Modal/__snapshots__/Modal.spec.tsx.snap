// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal matches snapshot 1`] = `
<div
  className="styles__ModalWrapper-sc-1gxqzoa-6 bfHzdf"
>
  <div
    className="styles__CardWrapper-sc-1gxqzoa-7 llgsuk"
  >
    <div
      className="styles__Card-sc-1gxqzoa-0 cPBKEa"
    >
      <header
        className="styles__Header-sc-1gxqzoa-1 knqvba"
      >
        <h1
          className="styles__Heading-sc-1gxqzoa-2 eqDSaw"
        >
          Discard unsaved changes?
        </h1>
        <svg
          aria-hidden="true"
          className="styles__CloseModalIcon-sc-1gxqzoa-8 cjHFEZ"
          focusable="false"
          height="15px"
          onClick={[Function]}
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </header>
    </div>
  </div>
  <div
    className="styles__Mask-sc-1gxqzoa-5 bsZwaD"
  />
</div>
`;

exports[`Modal matches snapshot with children 1`] = `
<div
  className="styles__ModalWrapper-sc-1gxqzoa-6 bfHzdf"
>
  <div
    className="styles__CardWrapper-sc-1gxqzoa-7 llgsuk"
  >
    <div
      className="styles__Card-sc-1gxqzoa-0 cPBKEa"
    >
      <header
        className="styles__Header-sc-1gxqzoa-1 knqvba"
      >
        <h1
          className="styles__Heading-sc-1gxqzoa-2 eqDSaw"
        >
          Discard unsaved changes?
        </h1>
        <svg
          aria-hidden="true"
          className="styles__CloseModalIcon-sc-1gxqzoa-8 cjHFEZ"
          focusable="false"
          height="15px"
          onClick={[Function]}
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </header>
      <div>
        child 1
      </div>
      <div>
        child 2
      </div>
    </div>
  </div>
  <div
    className="styles__Mask-sc-1gxqzoa-5 bsZwaD"
  />
</div>
`;
